### PR TITLE
Tablet extended sounds

### PR DESCRIPTION
**Pull request by vladest**
https://highfidelity.fogbugz.com/f/cases/6613/Add-Sound-Cues-to-OnHover-and-Click-Behavior

Test plan:
- open the tablet. Should have a sound emitted.
- Highlight over buttons, checkboxes, menu items, etc. Notice hover sounds
- Click on buttons, checkboxes, menu items, etc. Notice click sounds.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3388/HighFidelity-Beta-PR11387-e26118254fb5e70fe305c196215fe26e48c9aa0e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3388/HighFidelity-Beta-PR11387-e26118254fb5e70fe305c196215fe26e48c9aa0e.exe

**Comment by anshuman64** [comment]
@vladest Great start! Here's some feedback:

- Please add sounds to:
1. Highlight on radio buttons (Settings > General)
2. Highlight on buttons (Settings > General)
3. Click on buttons (Settings > General)
4. Highlight on keyboard 
5. Click on keyboard

- Stylus sound occurs when stylus enters tablet overlay. Instead, it should occur when the stylus becomes visible/invisible.

- Half-trigger laser causes permanent, annoying sound. Should not happen.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3586/HighFidelity-Beta-PR11387-1a2d38e31f681a5bf0bf24dde9a1c1be1bb942f2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3586/HighFidelity-Beta-PR11387-1a2d38e31f681a5bf0bf24dde9a1c1be1bb942f2.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3594/HighFidelity-Beta-PR11387-b190fac4f17668ff93f651cc7aa8a2a903b1f284.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3594/HighFidelity-Beta-PR11387-b190fac4f17668ff93f651cc7aa8a2a903b1f284.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3599/HighFidelity-Beta-PR11387-083b52abafb2cf4b7ac06d5a97ded2995ce0f677.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3599/HighFidelity-Beta-PR11387-083b52abafb2cf4b7ac06d5a97ded2995ce0f677.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3601/HighFidelity-Beta-PR11387-074362b22868c622f5b27943f07385d2cdb04cce.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3601/HighFidelity-Beta-PR11387-074362b22868c622f5b27943f07385d2cdb04cce.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3794/HighFidelity-Beta-PR11387-e360c1363e091bbc8ae8b0403a6712adc07b0d4a.dmg

**Comment by anshuman64** [comment]
@vladest Here is some design feedback:

- Remove sounds for stylus becoming visible and invisible
- Add highlight and click sounds to radio buttons in Audio menu
- menu>settings>edit: dividers and disabled buttons should not be highlighted or produce sound

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3991/HighFidelity-Beta-PR11387-e360c1363e091bbc8ae8b0403a6712adc07b0d4a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4083/HighFidelity-Beta-PR11387-2d010980b000f934db80ef7985f83fad1df2e259.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4083/HighFidelity-Beta-PR11387-2d010980b000f934db80ef7985f83fad1df2e259.exe

**Comment by Nex-Pro** [comment]
Testing this

**Comment by Nex-Pro** [comment]
### PASS on the test requirements.

open the tablet. Should have a sound emitted. **SUCCESS**
Highlight over buttons, checkboxes, menu items, etc. Notice hover sounds **SUCCESS**
Click on buttons, checkboxes, menu items, etc. Notice click sounds. **SUCCESS**

**My Feedback:**

Really brings the tablet interactions to life. +1

I concur with much of what @anshuman64 has said.
Did not get annoying sound on half laser.(blue)
I noticed that with very light trigger pressure it is possible to highlight the buttons before the lasers become active ! ( I kinda like that ).

More sounds on more elements.

A nice sound for when the tablet goes away please.

And of course option to customise every sound and to set the volume.

File: interface/resources/qml/controls-uit/Button.qml [hunk 1]
```diff
@@ -26,6 +27,16 @@ Original.Button {
 
     HifiConstants { id: hifi }
 
+    onHoveredChanged: {
+        if (hovered) {
+            tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/Button.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/Button.qml [hunk 2]
```diff
@@ -26,6 +27,16 @@ Original.Button {
...
+        }
+    }
+
+    onClicked: {
+        tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/Button.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/CheckBox.qml [hunk 3]
```diff
@@ -28,6 +30,15 @@ Original.CheckBox {
     readonly property int checkRadius: 2
     activeFocusOnPress: true
 
+    onClicked: {
+        tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/CheckBox.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/CheckBoxQQC2.qml [hunk 4]
```diff
@@ -32,6 +33,17 @@ CheckBox {
     readonly property int checkSize: Math.max(boxSize - 8, 10)
     readonly property int checkRadius: isRound ? checkSize / 2 : 2
     focusPolicy: Qt.ClickFocus
+    hoverEnabled: true
+
+    onClicked: {
+        tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/CheckBoxQQC2.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/CheckBoxQQC2.qml [hunk 5]
```diff
@@ -32,6 +33,17 @@ CheckBox {
...
+    }
+
+    onHoveredChanged: {
+        if (hovered) {
+            tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/CheckBoxQQC2.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/GlyphButton.qml [hunk 6]
```diff
@@ -24,6 +25,16 @@ Original.Button {
     width: 120
     height: 28
 
+    onHoveredChanged: {
+        if (hovered) {
+            tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/GlyphButton.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/GlyphButton.qml [hunk 7]
```diff
@@ -24,6 +25,16 @@ Original.Button {
...
+        }
+    }
+
+    onClicked: {
+        tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/GlyphButton.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/Key.qml [hunk 8]
```diff
@@ -32,8 +33,15 @@ Item {
             }
         }
 
+        onContainsMouseChanged: {
+            if (containsMouse) {
+                tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/Key.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/Key.qml [hunk 9]
```diff
@@ -32,8 +33,15 @@ Item {
...
+            }
+        }
+
         onClicked: {
             mouse.accepted = true;
+            tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/Key.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/controls-uit/RadioButton.qml [hunk 10]
```diff
@@ -27,6 +29,15 @@ Original.RadioButton {
     readonly property int checkSize: 10
     readonly property int checkRadius: 2
 
+    onClicked: {
+        tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/controls-uit/RadioButton.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/dialogs/preferences/ButtonPreference.qml [hunk 11]
```diff
@@ -22,7 +23,16 @@ Preference {
 
     Button {
         id: button
-        onClicked: preference.trigger()
+        onHoveredChanged: {
+            if (hovered) {
+                tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/dialogs/preferences/ButtonPreference.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/dialogs/preferences/ButtonPreference.qml [hunk 12]
```diff
@@ -22,7 +23,16 @@ Preference {
...
+            }
+        }
+
+        onClicked: {
+            preference.trigger()
```
Comment (ZappoMan) [inline interface/resources/qml/dialogs/preferences/ButtonPreference.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/dialogs/preferences/ButtonPreference.qml [hunk 13]
```diff
@@ -22,7 +23,16 @@ Preference {
...
+            tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/dialogs/preferences/ButtonPreference.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/dialogs/preferences/CheckBoxPreference.qml [hunk 14]
```diff
@@ -38,6 +39,16 @@ Preference {
 
     CheckBox {
         id: checkBox
+        onHoveredChanged: {
+            if (hovered) {
+                tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/dialogs/preferences/CheckBoxPreference.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/dialogs/preferences/CheckBoxPreference.qml [hunk 15]
```diff
@@ -38,6 +39,16 @@ Preference {
...
+            }
+        }
+
+        onClicked: {
+            tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/dialogs/preferences/CheckBoxPreference.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/hifi/audio/MicBar.qml [hunk 16]
```diff
@@ -57,8 +59,16 @@ Rectangle {
 
         hoverEnabled: true;
         scrollGestureEnabled: false;
-        onClicked: { Audio.muted = !Audio.muted; }
+        onClicked: {
+            Audio.muted = !Audio.muted;
+            tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/hifi/audio/MicBar.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/hifi/audio/MicBar.qml [hunk 17]
```diff
@@ -57,8 +59,16 @@ Rectangle {
...
+        }
         drag.target: dragTarget;
+        onContainsMouseChanged: {
+            if (containsMouse) {
+                tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/hifi/audio/MicBar.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/hifi/tablet/TabletButton.qml [hunk 18]
```diff
@@ -130,11 +131,13 @@ Item {
             }
             tabletButton.clicked();
             if (tabletRoot) {
-                tabletRoot.playButtonClickSound();
+                tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/hifi/tablet/TabletButton.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/hifi/tablet/TabletButton.qml [hunk 19]
```diff
@@ -130,11 +131,13 @@ Item {
...
             }
         }
         onEntered: {
             tabletButton.isEntered = true;
+            tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/hifi/tablet/TabletButton.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/hifi/tablet/TabletMenuView.qml [hunk 20]
```diff
@@ -69,12 +73,17 @@ FocusScope {
...
                 anchors.fill: parent
                 hoverEnabled: true
-                onEntered: listView.currentIndex = index
+                onEntered: {
+                    tabletInterface.playSound(TabletEnums.ButtonHover)
+                    listView.currentIndex = index
+                }
+
                 onClicked: {
+                    tabletInterface.playSound(TabletEnums.ButtonClick)
```
Comment (ZappoMan) [inline interface/resources/qml/hifi/tablet/TabletMenuView.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/hifi/tablet/TabletMenuView.qml [hunk 21]
```diff
@@ -69,12 +73,17 @@ FocusScope {
             onImplicitWidthChanged: listView !== null ? listView.recalcSize() : 0
 
             MouseArea {
+                enabled: name !== "" && item.enabled
                 anchors.fill: parent
                 hoverEnabled: true
-                onEntered: listView.currentIndex = index
+                onEntered: {
+                    tabletInterface.playSound(TabletEnums.ButtonHover)
```
Comment (ZappoMan) [inline interface/resources/qml/hifi/tablet/TabletMenuView.qml]: coding standard: lines should end with trailing `;`

File: interface/resources/qml/hifi/tablet/TabletMenuView.qml [hunk 22]
```diff
@@ -69,12 +73,17 @@ FocusScope {
...
+                    listView.currentIndex = index
```
Comment (ZappoMan) [inline interface/resources/qml/hifi/tablet/TabletMenuView.qml]: coding standard: lines should end with trailing `;`

File: libraries/ui/src/ui/TabletScriptingInterface.cpp [hunk 23]
```diff
@@ -63,6 +75,29 @@ TabletProxy* TabletScriptingInterface::getTablet(const QString& tabletId) {
...
+    }
+}
+
+void TabletScriptingInterface::playSound(TabletAudioEvents aEvent) {
+    SharedSoundPointer sound = _audioEvents[aEvent];
+    if (sound) {
+        AudioInjectorOptions options;
+        options.stereo = sound->isStereo();
+        options.ambisonic = sound->isAmbisonic();
+        options.localOnly = options.localOnly || sound->isAmbisonic();  // force localOnly when Ambisonic
```
Comment (ZappoMan) [inline libraries/ui/src/ui/TabletScriptingInterface.cpp]: hmmm -- why would we ever be playing these sounds as non-local only... that's pretty weird.

Where's the spec for this? This seems wrong.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4301/HighFidelity-Beta-PR11387-fa349a694c7c19f826f216f38d46661157cf4a8b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4301/HighFidelity-Beta-PR11387-fa349a694c7c19f826f216f38d46661157cf4a8b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4310/HighFidelity-Beta-PR11387-908d601a430b9da36cf0b1202750e5eec9e4b2d8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4310/HighFidelity-Beta-PR11387-908d601a430b9da36cf0b1202750e5eec9e4b2d8.exe

**Pull request merged**